Lib/os.py

```diff
@@ -416,14 +416,16 @@ def walk(top, topdown=True, onerror=None, followlinks=False):
             # Yield before sub-directory traversal if going top down
             yield top, dirs, nondirs
             # Traverse into sub-directories
-            for dirname in reversed(dirs):
-                new_path = join(top, dirname)
-                # bpo-23605: os.path.islink() is used instead of caching
-                # entry.is_symlink() result during the loop on os.scandir() because
-                # the caller can replace the directory entry during the "yield"
-                # above.
-                if followlinks or not islink(new_path):
-                    stack.append(new_path)
+            if dirs:
```

Comment:
How much of the gain is due to the elimination of `os.join` and how much due to the check on empty `dirs`? On my system a walk of a folder with github repositories (including the cpython repo) has `dirs` empty about 60% of the times.